drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+cache this information. (***TODO***: should have a reasonable 
+upper limit, e.g. 86400 seconds)
+
+***pointers***: a tuple with all the values this entry
+points towards, e.g. `{kind: "ipfs", data: "QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC"}`
+
+
+### Name
+
+Names MUST be normalized according to IDNA2008 before hashing, as 
+described in [uts-46](https://unicode.org/reports/tr46) and SHOULD 
+be stored in their ASCII representation.
+
+See also [rfc3491](https://tools.ietf.org/html/rfc3491) and 
+[rfc3492](https://tools.ietf.org/html/rfc3492).
+
+Names SHOULD be at most 253 characters long, in order to enable 
+interoperability with DNS.
+
+***Note:*** Fot this first draft names MUST have `.aet` or `.test`
+as a suffix, i.e. `mywallet.aet`.
+
+
+### Namespaces/Labels
+
+A name is split up into labels by the `FULL STOP (U+002E .)` character.
+These labels will also be referred to as namespaces, e.g. `mywallet.aet` 
+can be understood as the `mywallet` namespace in the `aet` namespace.
+
+A label MUST be longer than three characters.
+
+Labels SHOULD be at most 63 characters long and the full path
+of all names and namespaces SHOULD be at most 253 characters
+long, in order to enable interoperability with DNS.
+
+
+### Registrars
+
+A registrar controls who is allowed to claim names in their namespace
+and under what circumstances they are allowed to do so.
+
+The only available registrars for this AENS version will be hard-coded
+ones, which own the `.`, `.aet` and `.test` namespaces.
+
+The `.` namespace registrar is restricted and MUST NOT allow anyone to
+claim any names in its namespace.
+
+The `.aet` namespace registrar allows users to claim any name that is
+valid according to the validation rules for labels and names if they
+are the first to claim the name and pay the appropriate fee.
+
+The `.test` namespace registrar can be, as the name suggests, used for
+test purposes
+
+
+### Commitment Scheme
+
+Claiming a name requires a commitment scheme, which enables the claimant
+to commit to a name and after the commitment has been accepted into the 
+chain, reveal the name to finish the process.
+
+A commitment should be binding, i.e. the claimant cannot change
+the value they commited to withouth changing the actual commitment
+and hiding, so that a malicious miner learns nothing about the value
+the claimant has commited until they chose to reveal that value. This
+prevents malicious miners from front running, i.e. upon seeing a 
+claim transaction, including their own claim request for the same
+name instead of the original claimant's one.
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
```

Comment:
I think of the names as `subsub.sub.label.tld` wouldn't this mean that we should work recursively on the label? why is the recursion used anyway? 

`return sha256(namehash(label) + sha256(remainder))`

Reply:
Hashing the labels recursively preserves the hierarchical nature of the full name, i.e.

With the given definition we go from general to specific and yours does specific to general. I think either are ok.